### PR TITLE
Tyche final

### PR DESCRIPTION
**Pull request by ind-igo**

### CODE REVIEW

File: contracts/peripheral/YieldDirector.sol [hunk 1]
```diff
@@ -0,0 +1,364 @@
...
+        IERC20(sOHM).safeTransfer(msg.sender, amount_);
+
+        emit Withdrawn(msg.sender, recipient_, amount_);
+    }
+
+    /**
+        @notice Withdraw from all donor positions
+     */
+    function withdrawAll() external override {
+        require(disableWithdaws == false, "Withdraws currently disabled");
```
Comment (JExtor) [inline contracts/peripheral/YieldDirector.sol:160]: Could just do `!disableWithdaws`

File: contracts/peripheral/YieldDirector.sol [hunk 2]
```diff
@@ -0,0 +1,364 @@
...
+
+        DonationInfo[] storage donations = donationInfo[msg.sender];
+        require(donations.length != 0, "User not donating to anything");
+
+        uint256 sOhmIndex = IsOHM(sOHM).index();
+        uint256 total = 0;
```
Comment (JExtor) [inline contracts/peripheral/YieldDirector.sol:166]: can just do `uint256 total;`

File: contracts/peripheral/YieldDirector.sol [hunk 3]
```diff
@@ -0,0 +1,364 @@
...
+    }
+
+    /**
+        @notice Return total amount of donor's sOHM deposited
+     */
+    function totalDeposits(address donor_) external override view returns ( uint256 ) {
+        DonationInfo[] memory donations = donationInfo[donor_];
+        require(donations.length != 0, "User is not donating");
+
+        uint256 total = 0;
```
Comment (JExtor) [inline contracts/peripheral/YieldDirector.sol:211]: Could do `uint256 total;`

**Review by JExtor — COMMENTED** [review]
Looks pretty good. Just more polishing it seems to me. I didn't actually run through the math to make sure everything works as expected. But the logic seems pretty solid to me.

File: contracts/peripheral/YieldDirector.sol (hunk 2 above)
Comment (ind-igo) [reply contracts/peripheral/YieldDirector.sol:166]: Old habits 😅 I will keep though, I like the verbosity.

File: contracts/peripheral/YieldDirector.sol [hunk 4]
```diff
@@ -0,0 +1,364 @@
...
+    event Redeemed(address recipient_, uint256 amount_);
+    event EmergencyShutdown(bool active_);
+
+    constructor (address sOhm_) {
+        require(sOhm_ != address(0), "Invalid address for sOHM");
+
+        sOHM = sOhm_;
+        DECIMALS = ERC20(sOhm_).decimals();
+
+        disableDeposits = false;
```
[the remark this answers is not in the document]
Comment (ind-igo) [reply contracts/peripheral/YieldDirector.sol:60]: Ok noted, but I will leave in for clarity

File: contracts/peripheral/YieldDirector.sol [hunk 5]
```diff
@@ -0,0 +1,364 @@
...
+        return amount_
+            * (10 ** DECIMALS)
+            / (IsOHM(sOHM).index());
+    }
+
+    /**
+        @notice Convert agnostic value at current index to flat sOHM value
+        @dev Agnostic value earns rebases. Agnostic value is amount / rebase_index
+     */
+    function _fromAgnostic(uint256 amount_) internal view returns ( uint256 ) {
```
[the remark this answers is not in the document]
Comment (ind-igo) [reply contracts/peripheral/YieldDirector.sol:326]: True, but having an analog to `_toAgnostic` is probably good, no? Also, I need to replace with calls to sOHM anyway.

File: contracts/mocks/MockSOHM.sol [hunk 6]
```diff
@@ -0,0 +1,117 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.9;
+
+import {ERC20} from "@openzeppelin/contracts/token/ERC20/ERC20.sol";
+import "hardhat/console.sol";
+
+/**
+ * A mock version of sOHM, with an over-simplified rebase mechanism, for testing purposes only
+ */
+contract MockSOHM is ERC20 {
```
Comment (lolchocotaco) [inline contracts/mocks/MockSOHM.sol:10]: I love this alot.

File: contracts/peripheral/YieldDirector.sol [hunk 7]
```diff
@@ -0,0 +1,394 @@
...
+        @notice Deposit sOHM, records sender address and assign rebases to recipient
+        @param amount_ Amount of sOHM debt issued from donor to recipient
+        @param recipient_ Address to direct staking yield and vault shares to
+    */
+    function deposit(uint256 amount_, address recipient_) external override {
+        require(!depositDisabled, "Deposits currently disabled");
+        require(amount_ > 0, "Invalid deposit amount");
+        require(recipient_ != address(0), "Invalid recipient address");
+
+        IERC20(sOHM).safeTransferFrom(msg.sender, address(this), amount_);
```
Comment (lolchocotaco) [inline contracts/peripheral/YieldDirector.sol:74]: Would you want to enable gOHM deposits?

File: contracts/peripheral/YieldDirector.sol [hunk 8]
```diff
@@ -0,0 +1,394 @@
...
+
+    /**
+        @notice Get array index of a particular recipient in a donor's donationInfo array.
+        @return Array index of recipient address. If recipient not present, returns max uint256 value.
+     */
+    function _getRecipientIndex(address donor_, address recipient_) internal view returns (uint256) {
+        DonationInfo[] storage info = donationInfo[donor_];
+
+        uint256 existingIndex = MAX_UINT256;
+        for (uint256 i = 0; i < info.length; i++) {
```
Comment (lolchocotaco) [inline contracts/peripheral/YieldDirector.sol:330]: Any concerns around performance if this array is big af.

**Review by lolchocotaco — COMMENTED** [review]
I love the way things are laid out. 
The only concern I have is around the accounting for multiple donators to the same recipient across multiple epochs. 

Can you put in a test that accounts for this? 
Also might be able to do this via dapp tools as well. 

Additionally, should this be in another repo. 🤔

File: contracts/peripheral/YieldDirector.sol (hunk 7 above)
Comment (ind-igo) [reply contracts/peripheral/YieldDirector.sol:74]: > Can you put in a test that accounts for this? Also might be able to do this via dapp tools as well.

This is already taken care of

**Comment by ind-igo** [comment]
@lolchocotaco I need you ser. when you're back please help me merge this.

File: contracts/peripheral/YieldDirector.sol [hunk 9]
```diff
@@ -0,0 +1,390 @@
...
+
+        IERC20(sOHM).safeTransferFrom(msg.sender, address(this), amount_);
+
+        uint256 index = IsOHM(sOHM).index();
+
+        // Record donors's issued debt to recipient address
+        DonationInfo[] storage donations = donationInfo[msg.sender];
+        uint256 recipientIndex = _getRecipientIndex(msg.sender, recipient_);
+
+        if(recipientIndex == MAX_UINT256) {
```
Comment (Paul4912) [inline contracts/peripheral/YieldDirector.sol:82]: do you mean unequal? if this equal max you would not be able to push another item in correct?

Comment (Paul4912) [reply contracts/peripheral/YieldDirector.sol:82]: Not sure i understand the purpose of this

File: test/YieldDirector.test.js [hunk 10]
```diff
@@ -0,0 +1,501 @@
...
+
+    it('should set token addresses correctly', async () => {
+        await tyche.deployed();
+
+        expect(await tyche.sOHM()).to.equal(sOhm.address);
+    });
+
+    it('should deposit tokens to recipient correctly', async () => {
+        // Deposit 100 sOHM into Tyche and donate to Bob
+        const principal = "100000000000";
```
Comment (Paul4912) [inline test/YieldDirector.test.js:172]: think you should use ethers.utils.parseunits('100', 9) or write a function called toSOhmDecimals() that calls ethers.utils.parseunits('100', 9). Less mistakes not having to eyeball/count the number of zeros.

File: test/YieldDirector.test.js [hunk 11]
```diff
@@ -0,0 +1,501 @@
...
+        await tyche.deposit(principal, bob.address);
+
+        // Verify donor info
+        const donationInfo = await tyche.donationInfo(deployer.address, "0");
+        await expect(donationInfo.recipient).is.equal(bob.address);
+        await expect(donationInfo.deposit).is.equal(principal); // 10 * 10 ** 9
```
Comment (Paul4912) [inline test/YieldDirector.test.js:178]: comment should be 100 * 10 ** 9

File: test/YieldDirector.test.js [hunk 12]
```diff
@@ -0,0 +1,501 @@
...
+        //await expect(donationInfo.amount).is.equal(principal);
+
+        // Verify recipient data
+        const recipientInfo = await tyche.recipientInfo(bob.address);
+        await expect(recipientInfo.totalDebt).is.equal(principal);
+
+        const index = await sOhm.index();
+        await expect(recipientInfo.agnosticDebt).is.equal((principal / index) * 10 ** 9 );
```
Comment (Paul4912) [inline test/YieldDirector.test.js:186]: This looks weird to me, isn't principal a string from line 172?

File: contracts/peripheral/YieldDirector.sol [hunk 13]
```diff
@@ -0,0 +1,390 @@
...
+
+    /**
+        @notice Convert flat sOHM value to agnostic value at current index
+        @dev Agnostic value earns rebases. Agnostic value is amount / rebase_index.
+             1e9 is because sOHM has 9 decimals.
+     */
+    function _toAgnostic(uint256 amount_) internal view returns ( uint256 ) {
+        return amount_
+            * 1e9
+            / (IsOHM(sOHM).index());
```
Comment (Paul4912) [inline contracts/peripheral/YieldDirector.sol:343]: This is essentially converting the sOhm amount to gOhm amount right. gOhm has 18 decimals. Are you not concerned if the index becomes larger in future like let say 10000 by only using 9 decimals to store agnostic value we might lose some precision here?

Will probably be negligible amount but wanted to ask anyway.

**Comment by lolchocotaco** [comment]
Forgot to post this here:
Something about the block.timestamp is causing the tests to fail. Tests think that rebasing isn't happening. 

```10000000000 // Original Index
13950798 1641457375 // Blocknumber Timestamp
13950799 1642457375 // New blockNumber Timestamp
20010000000 // New Index 
13950800 1642457376 // BlockNumber timestamp
13950801 1643457376 // New BlockNumber Timestamp
20010000000 // New Index
13950802 1643457377
13950803 1644457377```

**Comment by Paul4912** [comment]
> Forgot to post this here: Something about the block.timestamp is causing the tests to fail. Tests think that rebasing isn't happening.
> 
> ```
> 13950798 1641457375 // Blocknumber Timestamp
> 13950799 1642457375 // New blockNumber Timestamp
> 20010000000 // New Index 
> 13950800 1642457376 // BlockNumber timestamp
> 13950801 1643457376 // New BlockNumber Timestamp
> 20010000000 // New Index
> 13950802 1643457377
> 13950803 1644457377```
> ```

I tried increasing timestamp in between each block so that its longer than 1 rebase(8 hours). Still getting same results. Might not be only issue.

**Review by Paul4912 — APPROVED** [review]

**Review by lolchocotaco — APPROVED** [review]
LGTM. lets goooooo

**Pull request merged**